controllers/heatcfnapi_controller.go

```diff
@@ -0,0 +1,662 @@
+			condition.UnknownCondition(condition.ServiceConfigReadyCondition, condition.InitReason, condition.ServiceConfigReadyInitMessage),
+			condition.UnknownCondition(condition.DeploymentReadyCondition, condition.InitReason, condition.DeploymentReadyInitMessage),
+		)
+
+		instance.Status.Conditions.Init(&cl)
+
+		if err := r.Status().Update(ctx, instance); err != nil {
+			return ctrl.Result{}, err
+		}
+	}
+	if instance.Status.Hash == nil {
+		instance.Status.Hash = map[string]string{}
+	}
+	if instance.Status.APIEndpoints == nil {
+		instance.Status.APIEndpoints = map[string]map[string]string{}
+	}
+	if instance.Status.ServiceIDs == nil {
+		instance.Status.ServiceIDs = map[string]string{}
+	}
+
+	helper, err := helper.NewHelper(
+		instance,
+		r.Client,
+		r.Kclient,
+		r.Scheme,
+		r.Log,
+	)
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+
+	defer func() {
+		if instance.IsReady() {
+			instance.Status.Conditions.MarkTrue(condition.ReadyCondition, condition.ReadyMessage)
+		}
+
+		if err := helper.SetAfter(instance); err != nil {
+			patch := client.MergeFrom(helper.GetBeforeObject())
+			if err := r.Status().Patch(ctx, instance, patch); err != nil && !k8s_errors.IsNotFound(err) {
+				util.LogErrorForObject(helper, err, "Update Status", instance)
+			}
+		}
+	}()
+
+	if !instance.DeletionTimestamp.IsZero() {
+		return r.reconcileDelete(ctx, instance, helper)
+	}
+
+	return r.reconcileNormal(ctx, instance, helper)
+}
+
+// SetupWithManager sets up the controller with the Manager.
+func (r *HeatCfnAPIReconciler) SetupWithManager(mgr ctrl.Manager) error {
+
+	configMapFn := func(o client.Object) []reconcile.Request {
+		result := []reconcile.Request{}
+
+		apis := &heatv1beta1.HeatCfnAPIList{}
+		listOpts := []client.ListOption{
+			client.InNamespace(o.GetNamespace()),
+		}
+		if err := r.Client.List(context.Background(), apis, listOpts...); err != nil {
+			r.Log.Error(err, "Unable to get CfnAPI CRs %v")
+			return nil
+		}
+
+		label := o.GetLabels()
+
+		if l, ok := label[labels.GetOwnerNameLabelSelector(labels.GetGroupLabel(heat.ServiceName))]; ok {
+			for _, cr := range apis.Items {
+				if l == heat.GetOwningHeatName(&cr) {
+					name := client.ObjectKey{
+						Namespace: o.GetNamespace(),
+						Name:      cr.Name,
+					}
+					r.Log.Info(fmt.Sprintf("ConfigMap object %s and CR %s marked with label: %s", o.GetName(), cr.Name, l))
+					result = append(result, reconcile.Request{NamespacedName: name})
+				}
+			}
+		}
+		if len(result) > 0 {
+			return result
+		}
+		return nil
+	}
+
+	return ctrl.NewControllerManagedBy(mgr).
+		For(&heatv1beta1.HeatCfnAPI{}).
+		Owns(&keystonev1.KeystoneService{}).
+		Owns(&keystonev1.KeystoneEndpoint{}).
+		Owns(&appsv1.Deployment{}).
+		Owns(&corev1.Secret{}).
+		Owns(&routev1.Route{}).
+		Owns(&corev1.Service{}).
+		// watch the config CMs we don't own
+		Watches(&source.Kind{Type: &corev1.ConfigMap{}},
+			handler.EnqueueRequestsFromMapFunc(configMapFn)).
```

Comment:
I wonder if we can consolidate this to just our Heat controller? I know we are doing it in engine and API, but Watches are expensive. Maybe we can trigger an action from the Heat controller and avoid having all of the other controllers also Watching the same resource?

I guess we can look into that in a seperate PR though. Just a thought as we add more controllers and more load on the API server.